[Windows] misc fixes to enable Windows build

SS-JIA opened the pull request:
Summary:
## Context

Implement some small miscellaneous fixes to enable installing ExecuTorch on Windows. Currently, the optimized kernels cannot be built successfully but the installation works in custom ops are not built.

## Installation

```powershell
cd executorch

# Need to disable building custom kernels; there are a bunch of build errors
$env:EXECUTORCH_BUILD_KERNELS_CUSTOM_AOT=0
# For some reason need to explicitly enable tensor extension otherwise I see a linker error
$env:CMAKE_ARGS="-DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON"

python install_executorch.py

# Unset if you want
Remove-Item Env:\EXECUTORCH_BUILD_KERNELS_CUSTOM_AOT
Remove-Item Env:\CMAKE_ARGS
```

## Testing

I am able to follow the [Getting Started tutorial](https://pytorch.org/executorch/stable/getting-started-setup.html#run-your-program) successfully.

## Outstanding Build Issues

Currently, with CMake I can successfully build with the following settings:

```
del -Recurse -Force cmake-out; `
  cmake . `
  -DCMAKE_INSTALL_PREFIX=cmake-out `
  -DPYTHON_EXECUTABLE=C:\\Users\\ssjia\\AppData\\Local\\miniconda3\\python.exe `
  -DCMAKE_PREFIX_PATH=C:\\Users\\ssjia\\AppData\\Local\\miniconda3\\Lib\\site-packages `
  -DCMAKE_BUILD_TYPE=Release `
  -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON `
  -DEXECUTORCH_BUILD_PYBIND=ON `
  -DEXECUTORCH_BUILD_XNNPACK=ON `
  -DEXECUTORCH_BUILD_KERNELS_QUANTIZED_AOT=ON `
  -DEXECUTORCH_BUILD_KERNELS_CUSTOM_AOT=ON `
  -DEXECUTORCH_BUILD_KERNELS_CUSTOM=OFF `
  -T ClangCL `
  -Bcmake-out; `
  cmake --build cmake-out -j64 --target install
```

If I switch `EXECUTORCH_BUILD_KERNELS_CUSTOM` to `ON`, then the build fails. The primary offenders appear to be the `gelu`. The implementation includes a header from ATen which is causing the majority of the errors.

SS-JIA commented on extension/llm/custom_ops/CMakeLists.txt:
```diff
@@ -49,7 +49,7 @@ if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(aarch64|arm64|armv7)$")
        "extension/llm/custom_ops/spinquant/third-party/FFHT/fht_neon.c"
   )
-elseif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "^(x86_64|AMD64)")
```
apparently AMD64 is synonymous with x86_64. On my Windows machine this value is AMD64.

SS-JIA commented on backends/xnnpack/CMakeLists.txt:
```diff
@@ -73,34 +73,42 @@ foreach(fbs_file ${_xnnpack_schema__srcs})
 )
 
-set(xnnpack_third_party pthreadpool cpuinfo)
+set(xnnpack_third_party pthreadpool extension_threadpool cpuinfo)
```
required due to [this recent refactor](https://github.com/pytorch/executorch/commit/6ba6e3ea6b2150e0fe2143b14e022be00f03dcb3)

SS-JIA replied:
without it, I see a linking error complaining that pthreadpool symbols cannot be found.

SS-JIA commented on backends/xnnpack/CMakeLists.txt:
```diff
@@ -73,34 +73,42 @@ foreach(fbs_file ${_xnnpack_schema__srcs})
 add_library(xnnpack_backend STATIC ${_xnnpack_backend__srcs})
 target_link_libraries(
-  xnnpack_backend PRIVATE ${xnnpack_third_party} executorch_core xnnpack_schema
+  xnnpack_backend PUBLIC ${xnnpack_third_party} executorch_core xnnpack_schema
```
changing to `PUBLIC` as suggested by @swolchok

SS-JIA commented on setup.py:
```diff
@@ -345,22 +345,28 @@ def inplace_dir(self, installer: "InstallerBuildExt") -> Path:
-                relative to the cmake-out directory. May be an fnmatch-style
-                glob that matches exactly one file. If the path ends in `.so`,
-                this class will also look for similarly-named `.dylib` files.
+            src_dir: The directory of the file to install, relative to the cmake-out
```
these changes are lifed from https://github.com/pytorch/executorch/pull/6979

tdakhran commented on backends/xnnpack/CMakeLists.txt:
```diff
@@ -73,34 +73,42 @@ foreach(fbs_file ${_xnnpack_schema__srcs})
 endforeach()
 
+if(WIN32)
+  set(MV_COMMAND powershell -Command "Move-Item -Path ${_xnnpack_flatbuffer__outputs} -Destination ${_xnnpack_schema__outputs}")
```
When cross-compiling for Windows, `WIN32` is set in cmake, but there is no powershell

SS-JIA replied:
good point. I may address this in another PR though. There are some other places in our CMake files that don't account for cross compiling for windows. So currently I prefer to address it in a follow up.

mcr229 replied:
do you mind adding this as a comment in this cmake? worthwhile to keep this info for the follow up next time.

SS-JIA replied:
yes, for sure.

SS-JIA commented on install_executorch.bat:
```diff
@@ -7,14 +7,8 @@ rem This batch file provides a basic functionality similar to the bash script.
-  )
-)
+rem Under windows, it's always python
+set PYTHON_EXECUTABLE=python
```
these changes are lifed from #6979

mcr229 reviewed (approved):
lgtm stamping to unblock

SS-JIA committed:
[Windows] misc fixes to enable Windows build

Summary:
## Context

Implement some small miscellaneous fixes to enable installing ExecuTorch on Windows. Currently, the optimized kernels cannot be built successfully but the installation works in custom ops are not built.

## Installation

```powershell
cd executorch

# Need to disable building custom kernels; there are a bunch of build errors
$env:EXECUTORCH_BUILD_KERNELS_CUSTOM_AOT=0
# For some reason need to explicitly enable tensor extension otherwise I see a linker error
$env:CMAKE_ARGS="-DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON"

python install_executorch.py

# Unset if you want
Remove-Item Env:\EXECUTORCH_BUILD_KERNELS_CUSTOM_AOT
Remove-Item Env:\CMAKE_ARGS
```

## Testing

I am able to follow the [Getting Started tutorial](https://pytorch.org/executorch/stable/getting-started-setup.html#run-your-program) successfully.

## Outstanding Build Issues

Currently, with CMake I can successfully build with the following settings:

```
del -Recurse -Force cmake-out; `
  cmake . `
  -DCMAKE_INSTALL_PREFIX=cmake-out `
  -DPYTHON_EXECUTABLE=C:\\Users\\ssjia\\AppData\\Local\\miniconda3\\python.exe `
  -DCMAKE_PREFIX_PATH=C:\\Users\\ssjia\\AppData\\Local\\miniconda3\\Lib\\site-packages `
  -DCMAKE_BUILD_TYPE=Release `
  -DEXECUTORCH_BUILD_EXTENSION_TENSOR=ON `
  -DEXECUTORCH_BUILD_PYBIND=ON `
  -DEXECUTORCH_BUILD_XNNPACK=ON `
  -DEXECUTORCH_BUILD_KERNELS_QUANTIZED_AOT=ON `
  -DEXECUTORCH_BUILD_KERNELS_CUSTOM_AOT=ON `
  -DEXECUTORCH_BUILD_KERNELS_CUSTOM=OFF `
  -T ClangCL `
  -Bcmake-out; `
  cmake --build cmake-out -j64 --target install
```

If I switch `EXECUTORCH_BUILD_KERNELS_CUSTOM` to `ON`, then the build fails. The primary offenders appear to be the `gelu`. The implementation includes a header from ATen which is causing the majority of the errors.

SS-JIA commented on CMakeLists.txt:
```diff
@@ -506,13 +506,18 @@ if(EXECUTORCH_BUILD_FLATC)
+  if(WIN32)
+    # flatbuffers does not use CMAKE_BUILD_TYPE. Internally, the build forces Release
+    # config, but from CMake's perspective the build type is always Debug.
+    set(FLATC_EXECUTABLE ${BINARY_DIR}/$<CONFIG>/flatc.exe)
```
@jathu this is the change that was needed

jathu reviewed (approved):
flatc changes lgtm
